hardware/comm.py

```diff
@@ -0,0 +1,126 @@
+    
+        R_interp = np.linspace(430,10000,1000) ## MR -- finer steps of R
+        Pt_error_intp = interp1d(self.R,self.Pt_error)  ## MR -- interp function
+    
+        Off_pt_angle = 0 ## MR -- ideal case of directly overhead? Variation in position angle
+        Pt_error_main = Pt_error_intp(Srange) ## MR -- pt 
+    
+        Antenna_gain_intp = np.linspace(0,10,1000) 
+        SANT_intp = interp1d(self.Antenna_gain,self.SANT,fill_value="extrapolate")
+        #print(Off_pt_angle + Pt_error_main)
+        SANT_main = SANT_intp(Off_pt_angle+Pt_error_main)
+        Antenna_return_loss = 15 ## MR-- E loss due to refections
+        Mismatch_loss = 10*np.log10(1-(10**(-Antenna_return_loss/20))**2) ## MR --impedance mismatch? 
+        SC_noise_temp = 26.8
+        SCGT = SANT_main + Mismatch_loss - SC_noise_temp 
+        
+        Uplink_CN0 = Asset_EIRP + free_space_path_loss + SCGT - 10*np.log10(1.38e-23) ## MR --system carrier to noise
+        Mod_loss = 0.0  #says calculated but given
+        Implementation_loss = -1.0  #assumed
+        Pll_bw_Hz = 700 #assumed ## MR-- Phase locked loop
+        Pll_bw_dB = 10*np.log10(Pll_bw_Hz)
+        SN_loop = 17.9470058901322 ## MR -- SNR
+    
+        Carrier_margin = Uplink_CN0 + Implementation_loss - Pll_bw_dB - SN_loop 
+
+        #rate_pw2 = self.max_rate_kbps ## MR -- 1024 so 2**32? is this the max rate??
+        #print('Rate is',rate_pw2) ## MR -- debug 
+        
+        rate_pw2 = np.sqrt(max_rate_kbps)
+        Coded_symb_rt_input = rate_pw2 ## MR -- symbols transmitted per second
+        Coded_symb_rt = 2**Coded_symb_rt_input
+    
+        Code_rate = 0.662430862918876   #theory ## MR -- ratio of useful data bits to total transmitted bits
+        Data_rate = Coded_symb_rt * Code_rate ## MR -- data rate?
+    
+        EbN0 = Uplink_CN0 + Implementation_loss - 10*np.log10(Data_rate*1000) ## MR -- E per bit --> noise power spec density ratio?
+    
+        Threshold_EbN0 = 2.1
+        Data_demod_margin = EbN0 - Threshold_EbN0
+        
+        return Data_demod_margin, Data_rate
+        
+
+    def get_rate(self, distance_km, alt_deg, demod_marg, zero_ext_gain=False):
+        '''
+        Data transfer rate calculation. Begin with minimum power 2**5, tests out different powers and different 
+        demods and rates. Converges when maximum pw2 reached (12) or minimum demod_marg (3) is acheived.
+        
+        input: distance (array), alt_deg (array), demod_marg (integer), zero_ext_gain (boolean)
+        output: rate (in what units?), demod, bit rate?
+        
+        '''
+        extra_gain = self.ext_gain(alt_deg) if not zero_ext_gain else 0
+        pw2 = 5
+        demod_marg = self.link_margin_dB
+        print('Demod margin is',demod_marg)
+        while True:
+            try_demod, try_rate = self.demodulation(distance_km, pw2, extra_gain)
+            if (try_demod < demod_marg) or (pw2 == 12):
```

Comment:
pw2 == int(np.log10(max_rate_kbps)/np.log10(2))-1

We need the -1 to prevent the next iteration (pw2 +=1) from going over the max of 10. Also the (2**(pw2-1)) in the output of the function has -1 to do the same thing.